packages/gatsby-theme-docs/src/plugins/rehype-id-slug.js

```diff
@@ -0,0 +1,26 @@
+const visit = require('unist-util-visit');
+const hasProperty = require('hast-util-has-property');
+const headingRank = require('hast-util-heading-rank');
+const toString = require('hast-util-to-string');
+const Slugger = require('github-slugger');
+
+const slugs = new Slugger();
+
+/**
+ * This plugins is a re-implementation of rehype-slug https://github.com/rehypejs/rehype-slug
+ * The purpose of the plugin is to add id to headings based on the text content.
+ * This re-implementation simply ensure that the ids don't contain trailing dashes or underscores, the
+ * rest of the implementation works exactly like the original.
+ *
+ * TODO: ehnancement: add the possibility of passing in a 'options.postProcess' function to be able to sanitize
+ * the slugs with a custom function implementation
+ */
+module.exports = () => (tree) => {
+  slugs.reset();
+  visit(tree, 'element', (node) => {
+    if (headingRank(node) && node.properties && !hasProperty(node, 'id')) {
+      const nodeSlug = slugs.slug(toString(node));
+      node.properties.id = nodeSlug.replace(/[-_]+$/, '');
```

Comment:
The intent is not to remove them completely, the problem we face in various occasions is that there are trailing or leading slashes / underscores.   the root-rootcause is that the slugger does have issues with MDX components in the content. 

Didn't you work with a function that gets the pure text content from a piece of MdX in some other place? 

Ideally this implementation is re-using a function that is used in the matching implementation that generates the ToC

Reply:
The AST node structure is something like this:
```
node {
  type: 'element',
  tagName: 'h1',
  properties: {},
  children: [
    {
      type: 'text',
      value: 'Get Customer in Store ',
      position: [Object]
    },
    {
      type: 'mdxJsxTextElement',
      name: 'Beta',
      attributes: [],
      children: [],
      position: [Object],
      data: [Object]
    }
  ],
  position: {
    start: { line: 7, column: 1, offset: 29 },
    end: { line: 7, column: 33, offset: 61 }
  }
}
```
and the `toString` implementation is basically concatenating all the text nodes in a single string that gets passed to the `slug` function.
As you can see, the empty space is in the text node, and it gets converted to `-`. So this is how it works, I'm  not sure what you mean with `the slugger does have issues with MDX components in the content`. The slugger doesn't even see the mdx components.

Reply:
Yes, you're right.  I was still mentally on the MDX v1 implementation which did not have as proper an AST.